beacon-chain/db/kv/wss.go

```diff
@@ -7,44 +7,46 @@ import (
 
 	"github.com/pkg/errors"
 	types "github.com/prysmaticlabs/eth2-types"
-	statev2 "github.com/prysmaticlabs/prysm/beacon-chain/state/v2"
 	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/proto/detect"
 	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
-	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/wrapper"
 )
 
 // SaveOrigin loads an ssz serialized Block & BeaconState from an io.Reader
 // (ex: an open file) prepares the database so that the beacon node can begin
 // syncing, using the provided values as their point of origin. This is an alternative
 // to syncing from genesis, and should only be run on an empty database.
 func (s *Store) SaveOrigin(ctx context.Context, stateReader, blockReader io.Reader) error {
-	// unmarshal both block and state before trying to save anything
-	// so that we fail early if there is any issue with the ssz data
-	blk := &ethpb.SignedBeaconBlockAltair{}
+	sb, err := ioutil.ReadAll(stateReader)
+	if err != nil {
+		return errors.Wrap(err, "failed to read origin state bytes")
+	}
 	bb, err := ioutil.ReadAll(blockReader)
 	if err != nil {
 		return errors.Wrap(err, "error reading block given to SaveOrigin")
 	}
-	if err := blk.UnmarshalSSZ(bb); err != nil {
-		return errors.Wrap(err, "could not unmarshal checkpoint block")
+
+	cf, err := detect.ByState(sb)
```

Comment:
Thinking about this API...if I were looking for a tool to accomplish this and came across this package, I would expect it to be named more like:
```go
versionDecoder := detect.FromBeaconStateBytes(sb)
wrappedBlock := versionDecoder.UnmarshalBeaconBlock(bb)
```
Alternatively, an approach that could make things more idiomatic would be to define some kind of VersionedDecoder interface and you can do something like `decoder := versionedDecoder.New(stateReader)` which takes in an io.Reader, and then you can have a set of common utilities that take in the VersionedDecoder as an interface. An analogy for this is the JSONDecoder, which exposes a Decode method. I can foresee us using this utility across the repo in various places